torch/lib/c10d/test/TCPStoreTest.cpp

```diff
@@ -0,0 +1,87 @@
+#include "StoreTestCommon.hpp"
+#include "TcpStore.hpp"
+
+#include <thread>
+#include <iostream>
+#include <cstdlib>
+
+
+int main(int argc, char** argv) {
+
+  // Master store
+  c10d::TCPStore masterStore("127.0.0.1", 29500, true);
+
+  // Basic set/get on the master store
+  c10d::test::set(masterStore, "key0", "value0");
+  c10d::test::set(masterStore, "key1", "value1");
+  c10d::test::set(masterStore, "key2", "value2");
+  c10d::test::check(masterStore, "key0", "value0");
+  c10d::test::check(masterStore, "key1", "value1");
+  c10d::test::check(masterStore, "key2", "value2");
+
+  // Hammer on TCPStore
+  std::vector<std::thread> threads;
+  const auto numThreads = 16;
+  const auto numIterations = 1000;
+  c10d::test::Semaphore sem1, sem2;
+
+  // Each thread will have a slave store to send/recv data
+  std::vector<std::unique_ptr<c10d::TCPStore>> slaveStores;
+  for (auto i = 0; i < numThreads; i++) {
+    slaveStores.push_back(std::unique_ptr<c10d::TCPStore>(new
+      c10d::TCPStore("127.0.0.1", 29500, false)));
+  }
+
+  for (auto i = 0; i < numThreads; i++) {
+    threads.push_back(std::move(std::thread([&sem1, &sem2, &slaveStores, i] {
+            for (auto j = 0; j < numIterations; j++) {
+              slaveStores[i]->add("counter", 1);
+            }
+            // Let each thread set and get key on its slave store
+            std::string key = "thread_" + std::to_string(i);
+            for (auto j = 0; j < numIterations; j++) {
+              std::string val = "thread_val_" + std::to_string(j);
+              c10d::test::set(*slaveStores[i], key, val);
+              c10d::test::check(*slaveStores[i], key, val);
+            }
+
+            sem1.post();
+            sem2.wait();
+
+            // Now check other threads' written data
+            for (auto j = 0; j < numThreads; j++) {
+              if (j == i) {
+                continue;
+              }
+              std::string key = "thread_" + std::to_string(i);
+              std::string val = "thread_val_" +
+                  std::to_string(numIterations - 1);
+              c10d::test::check(*slaveStores[i], key, val);
+            }
```

Comment:
Can you also check the counter in all threads?

Reply:
done